physlr/physlr.py

```diff
@@ -1332,6 +1332,79 @@ def determine_molecules_bc_cosine_of_squared(g, u):
+        """Prtition the subgraph into bins randomly for faster processing. Return bins."""
+        bins_count = 1
+        if len(node_set) > max_size:
+            bins_count = 1 + int(len(node_set) / max_size)
+        node_list = list(node_set)
+        random.shuffle(node_list)
+        size, leftover = divmod(len(node_set), bins_count)
+        bins = [node_list[0 + size * i: size * (i + 1)] for i in list(range(bins_count))]
+        edge = size * bins_count
+        for i in list(range(leftover)):
+            bins[i % bins_count].append(node_list[edge + i])
+        bin_sets = [set() for _ in range(len(bins))]
+        for i, c in zip(range(len(bins)), bins):
+            bin_sets[i].update(set(c))
+        return bin_sets
+
+    @staticmethod
+    def merge_communities(g, communities, node_set=0, strategy=0, cutoff=20):
+        """Merge communities if appropriate."""
+        mode = 1
+        if cutoff == -1:  # no merging
+            return communities
+        if len(communities) == 1 and (node_set == 0 or strategy != 1):
+            return communities
+        if strategy == 1:  # Merge by Initializing Louvain with the communities
+            return Physlr.detect_communities_louvain(g, node_set, communities)
+        # Ad-hoc Merge (default - strategy = 0)
+        merge_network = nx.Graph()
+        for i in range(len(communities)):
+            merge_network.add_node(i)
+        for i, com1 in enumerate(communities):
+            for k, com2 in enumerate(communities):
+                if i < k:
+                    if mode == 1:  # disjoint input communities.
+                        if nx.number_of_edges(
+                                g.subgraph(com1.union(com2))) - \
+                                nx.number_of_edges(g.subgraph(com1)) - \
+                                nx.number_of_edges(g.subgraph(com2)) > cutoff:
+                            merge_network.add_edge(i, k)
+                    else:  # overlapping input communities.
+                        if nx.number_of_edges(
+                                g.subgraph(com1.union(com2))) - \
+                                len(set(g.subgraph(com1).edges()).union(
+                                    set(g.subgraph(com2).edges()))) \
+                                > cutoff:
+                            merge_network.add_edge(i, k)
+        return [{barcode for j in i for barcode in communities[j]}
+                for i in nx.connected_components(merge_network)]
+
+    @staticmethod
+    def determine_molecules_partition_split_merge(g, u):
+        """
+        Assign the neighbours of this vertex to molecules using fast heuristic community detection.
+        Pipeline: bi-connected (bc) + partition + bc + k-cliques communities + merge.
+        """
+        return [merged
+                for bi_connected_component in
+                Physlr.detect_communities_biconnected_components(g, set(g.neighbors(u)))
```

Comment:
```suggestion
                Physlr.detect_communities_biconnected_components(g, g.neighbors(u))
```
Is this conversion necessary?

Reply:
all `detect_communities_*` functions accept a node set. I am checking if I can change it to list of list... but on top of my head we needed it for mixing these functions. I might be wrong. I 'll check.

Reply:
yes it's necessary.
we have `components = list(nx.connected_components(g.subgraph(node_set - cut_vertices)))` in `components = list(nx.connected_components(g.subgraph(node_set - cut_vertices)))` which requires `node_set` to be a set. We have kind of same thing in other com-det methods. Any alternatives for it?

Reply:
`g[u].keys()` may be more efficient than `set(g.neighbors(u))`. Want to time it?

Reply:
woof really different:
```
print(min(timeit.Timer('a = g[u].keys()', setup=setup).repeat(7, 1000000)))
0.04027111700270325
print(min(timeit.Timer('a = set(g.neighbors(u))', setup=setup).repeat(7, 1000000)))
0.6511558880010853
```
tested it for a node with 12 neighbors.

Reply:
changed and tested

Reply:
Nice! That's a win!